File: 1894-merge-strings-alternately/merge-strings-alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        result = []
        i = 0  # pointer for word1
        j = 0  # pointer for word2

        # Alternate taking characters from both strings
        while i < len(word1) and j < len(word2):
            result.append(word1[i])
            result.append(word2[j])
            i += 1
            j += 1

        # Append remaining characters from word1 (if any)
        while i < len(word1):
            result.append(word1[i])
            i += 1

        # Append remaining characters from word2 (if any)
        while j < len(word2):
            result.append(word2[j])
            j += 1

        return ''.join(result)
        

File: 1894-merge-strings-alternately/merge-strings-alternately.cpp
class Solution {
public:
    string mergeAlternately(string word1, string word2) {
        string result;
        result.reserve(word1.size() + word2.size());  // avoid reallocations
        
        int i = 0, j = 0;

        // Alternate taking characters from both strings
        while (i < word1.size() && j < word2.size()) {
            result += word1[i];
            result += word2[j];
            i++;
            j++;
        }

        // Append remaining characters from word1
        while (i < word1.size()) {
            result += word1[i];
            i++;
        }

        // Append remaining characters from word2
        while (j < word2.size()) {
            result += word2[j];
            j++;
        }

        return result;
    }
};